Automated Smoketest: :x:
OS: centos 7.6.1810 (Linux 3.10.0-957.1.3.el7.x86_64)
 GCC: gcc (GCC) 7.3.1 20180303 (Red Hat 7.3.1-5)
 Host: ip-10-20-0-135.ca-central-1.compute.internal
Sha: 2cb11081f00665d4ed87563ac0ffb243d5dd4ab2
Build: success
ECL Watch:  Rebuilding Site
Install error(s): 
npm ERR! code E409
npm ERR! 409 Conflict - GET https://registry.npmjs.org/@hpcc-js/chart/-/chart-2.19.1.tgz
npm ERR! A complete log of this run can be found in:
npm ERR!     /home/centos/.npm/_logs/2020-04-06T14_07_44_523Z-debug.log

Install warning(s): 
npm WARN optional SKIPPING OPTIONAL DEPENDENCY: fsevents@1.2.9 (node_modules/watchpack/node_modules/fsevents):

Lint error(s): 
res:
> eclwatch@1.0.0 test /home/centos/smoketest/PR-13597/HPCC-Platform/esp/src
> run-s lint
sh: run-s: command not found
npm ERR! Test failed.  See above for more details.

Milestone:Install hpccsystems-platform-community_7.6.37-closedown0.el7.x86_64.rpm
Installed (2020-04-06_14-18-16 23 sec )
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | elaps | 
|---|---|---|---|---|---|---|
|unittest |116 |116 |0 |0 |0 |40 sec |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |1 sec |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |8 sec |



Regression test result:

| phase | total | pass | fail | elaps | 
|---|---|---|---|---|
|setup (hthor) |11 |11 |0 |23 sec  (00:00:23) |
|setup (thor) |11 |11 |0 |44 sec  (00:00:44) |
|setup (roxie) |11 |11 |0 |22 sec  (00:00:22) |
|test (hthor) |869 |869 |0 |303 sec  (00:05:03) |
|test (thor) |808 |808 |0 |703 sec  (00:11:43) |
|test (roxie) |950 |950 |0 |375 sec  (00:06:15) |


HPCC Stop: OK
HPCC Uninstall: OK


 ... ( comment is too long, 61 lines, 2242 bytes, truncated) ... 
[Build error:](https://gist.github.com/f7e856c01e6d9f26115a18b936782b1a)
Build log [Build log](https://gist.github.com/HPCCSmoketest/f7e856c01e6d9f26115a18b936782b1a#file-result-20-04-06-14-07-02-log)

Time stats:

| Prep time | Build time | Package time | Install time | Start time | Test time | Stop time | Summary | 
|---|---|---|---|---|---|---|---|
|    9 sec (00:00:09) |  409 sec (00:06:49) |  198 sec (00:03:18) |   23 sec (00:00:23) |   19 sec (00:00:19) | 1694 sec (00:28:14) |   17 sec (00:00:17) | 2369 sec (00:39:29) |

```

--- a/tools/testsocket/testsocket.cpp
+++ b/tools/testsocket/testsocket.cpp
@@ -74,6 +74,10 @@ Semaphore finishedReading;
 FILE * trace;
 CriticalSection traceCrit;
 
+unsigned queryDelayMS = 0;
+unsigned totalQueryCnt = 0;
+double totalQueryMS = 0.0;
+
 //---------------------------------------------------------------------------
 
 void SplitIpPort(StringAttr & ip, unsigned & port, const char * address)
@@ -513,6 +517,10 @@ int doSendQuery(const char * ip, unsigned port, const char * base)
     Owned<ISecureSocketContext> secureContext;
     __int64 starttime, endtime;
     StringBuffer ipstr;
+    CTimeMon tm;
+    if (queryDelayMS)
+        tm.reset(queryDelayMS);
+
     try
     {
         if (strcmp(ip, ".")==0)
@@ -733,9 +741,12 @@ int doSendQuery(const char * ip, unsigned port, const char * base)
                 fprintf(trace, "%s", result.str());
             }
 
+            double queryTimeMS = (double)(cycle_to_nanosec(endtime - starttime))/1000000;
+            totalQueryMS += queryTimeMS;
+            totalQueryCnt++;
             if (showTiming && rawOnly == false)
             {
-                fprintf(trace, "Time taken = %.3f msecs\n", (double)(cycle_to_nanosec(endtime - starttime)/1000000));
+                fprintf(trace, "Time taken = %.3f msecs\n", queryTimeMS);
                 fputs("----------------------------------------------------------------------------\n", trace);
             }
         }
@@ -745,6 +756,14 @@ int doSendQuery(const char * ip, unsigned port, const char * base)
     {
         socket->close();
     }
+
+    if (queryDelayMS)
+    {
+        unsigned remaining;
+        if (!tm.timedout(&remaining))
+            Sleep(remaining);
+    }
+
     return 0;
 }
 
@@ -758,7 +777,7 @@ public:
 
 protected:
     StringAttr      ip;
-    unsigned            port;
+    unsigned        port;
     StringAttr      base;
 };
 
@@ -799,6 +818,7 @@ void usage(int exitCode)
     printf("  -pr <text>add a prefix to the query\n");
     printf("  -q        quiet - don't echo query\n");
     printf("  -qname xx Use xx as queryname in place of the xml root element name\n");
+    printf("  -qd       delay (ms) to possibly wait between start of query and start of next query\n");
     printf("  -r <n>    repeat the query several times\n");
     printf("  -rl       roxie logfile mode\n");
     printf("  -rs       remote stream request\n");
@@ -950,6 +970,14 @@ int main(int argc, char **argv)
         else if (stricmp(argv[arg], "-q") == 0)
         {
             echoSingle = false;
+            ++arg;
+        }
+        else if (stricmp(argv[arg], "-qd") == 0)
+        {
+            ++arg;
+            if (arg>=argc)
+                usage(1);
+            queryDelayMS = atoi(argv[arg]);
             ++arg;
         }
         else if (stricmp(argv[arg], "-qname") == 0)
@@ -1173,7 +1201,12 @@ int main(int argc, char **argv)
         {
             if (trace != NULL)
             {
-                fprintf(trace, "Total Time taken = %.3f msecs\n", (double)(cycle_to_nanosec(endtime - starttime)/1000000));
+                fprintf(trace, "Total Time taken = %.3f msecs\n", (double)(cycle_to_nanosec(endtime - starttime))/1000000);
+                if (totalQueryCnt)
+                {
+                    double timePerQueryMS = totalQueryMS / totalQueryCnt;
+                    fprintf(trace, "Total Queries: %u Avg t/q = %.3f msecs\n", totalQueryCnt, timePerQueryMS);
+                }
                 fputs("----------------------------------------------------------------------------\n", trace);
             }
         }

